R/params.R

```diff
@@ -0,0 +1,329 @@
+    }
+  }
+  
+  # merge explicitly provided params with defaults
+  merge_lists(default_params, params, recursive = FALSE)
+}
+
+params_converters <- list(
+    logical = function(value) {
+      as.logical(value)
+    },
+    Date = function(value) {
+      as.Date(value)
+    },
+    POSIXct = function(value) {
+      ## FIXME: This is how knitr does datetime conversion, but maybe we should use the local tz.
+      as.POSIXct(value, tz="GMT")
+    },
+    integer = function(value) {
+      as.integer(value)
+    },
+    numeric = function(value) {
+      as.numeric(value)
+    },
+    character = function(value) {
+      as.character(value)
+    },
+    data.frame = function(value) {
+      read.csv(value$datapath)
+    }
+    )
+
+params_get_converter <- function(param) {
+  for (c in param$class) {
+    converter <- params_converters[[c]]
+    if (!is.null(converter)) {
+      return(converter)
+    }
+  }
+  ## FIXME: this probably should be an error, as it is an unsupported type.
+  return(function(value){ value })
+}
+
+params_controls <- list()
+params_controls$integer <- params_controls$numeric <- function(param) {
+  ## If min/max are specified, use a slider.
+  if (is.null(param$min) || is.null(param$max)) {
+    shiny::numericInput
+  } else {
+    shiny::sliderInput
+  }
+}
+
+params_controls$logical <- function(param) { shiny::checkboxInput }
+## BUG: dateInput does not allow the user to not specify a value.
+##     https://github.com/rstudio/shiny/issues/896
+params_controls$Date <- function(param) { shiny::dateInput }
+## BUG: shiny does not support datetime selectors
+##     https://github.com/rstudio/shiny/issues/897
+##     we ask for string input for now.
+params_controls$POSIXct <- function(param) { shiny::textInput }
+params_controls$character <- function(param) { shiny::textInput }
+params_controls$data.frame <- function(param) { shiny::fileInput }
+params_get_control <- function(param) {
+  ## A value might have multiple classes. Try: class(Sys.time())
+  ## Try to find first class listed with a named control.
+  for (c in param$class) {
+    control <- params_controls[[c]]
+    if (!is.null(control)) {
+      return(control(param))
+    }
+  }
+  return(NULL)
+}
+
+params_configurable <- function(param) {
+  (!is.null(params_get_control(param))) && ( length(param$value) == 1 || "data.frame" %in% param$class )
+}
+
+#' Run a shiny application asking for parameter configuration for the given document.
+#'
+#' @param file Path to the R Markdown document with configurable parameters.
+#' @param input_lines Content of the R Markdown document. If \code{NULL}, the contents of \code{file} will be read.
+#' @param params A named list of optional parameter overrides used in place of the document defaults.
+#' @param shiny_args Additional arguments to \code{\link[shiny:runApp]{runApp}}.
+#'
+#' @return named list with overridden parameter names and value.
+#' 
+#' @export
+knit_params_ask <- function(file = "index.Rmd",
```

Comment:
I don't think we should default the `file` argument (there's no strong convention around using `index.Rmd` as a filename so this will seem random, less confusing to force the specification I think).